cxx/gendb.cc

```diff
@@ -0,0 +1,157 @@
+  for (const auto& [query_rel, val] : vals) {
+    // Sample a set of items to be incorporated into the query relation.
+    const std::vector<std::string>& class_path =
+        schema.query.fields.at(query_rel).class_path;
+    std::vector<int> items =
+        sample_entities_relation(prng, schema.query.record_class,
+                                 class_path.cbegin(), class_path.cend(), id);
+
+    // Incorporate the items/value into the query relation.
+    incorporate_query_relation(prng, query_rel, items, val);
+  }
+}
+
+// This function walks the class_path of the query, populates the global
+// reference_values table if necessary, and returns a sampled set of items
+// for the query relation.
+std::vector<int> GenDB::sample_entities_relation(
+    std::mt19937* prng, const std::string& class_name,
+    std::vector<std::string>::const_iterator class_path_start,
+    std::vector<std::string>::const_iterator class_path_end, int class_item) {
+  if (class_path_end - class_path_start == 1) {
+    // These are domains and we need to DFS-traverse the class's
+    // parents, similar to PCleanSchemaHelper::compute_domains_for.
+    return sample_class_ancestors(prng, class_name, class_item);
+  } else {
+    // These are noisy relation domains along the path from the latent cleanly-
+    // observed class to the record class.
+    std::string ref_field = *class_path_start;
+
+    // If the reference field isn't populated, sample a value from a CRP and
+    // add it to reference_values.
+    std::string ref_class =
+        std::get<ClassVar>(
+            schema.classes.at(class_name).vars.at(ref_field).spec)
+            .class_name;
+    if (!reference_values[class_name].contains(class_item)) {
+      sample_and_incorporate_reference(prng, class_name, class_item, ref_field,
+                                       ref_class);
+    }
+    std::vector<int> items = sample_entities_relation(
+        prng, ref_class, ++class_path_start, class_path_end,
+        reference_values[class_name][class_item][ref_field]);
+    items.push_back(class_item);
+    return items;
+  }
+}
+
+void GenDB::sample_and_incorporate_reference(std::mt19937* prng,
+                                             const std::string& class_name,
+                                             int class_item,
+                                             const std::string& ref_field,
+                                             const std::string& ref_class) {
+  int new_val = domain_crps[ref_class].sample(prng);
+
+  // Generate a unique ID for the sample and incorporate it into the
+  // domain CRP.
+  std::stringstream new_id_str;
+  new_id_str << class_name << class_item << ref_field;
+  int new_id = std::hash<std::string>{}(new_id_str.str());
+  reference_values[class_name][class_item][ref_field] = new_val;
+  domain_crps[ref_class].incorporate(new_id, new_val);
+}
+
+// Recursively incorporates samples into base relations.
```

Comment:
Please loudly comment that query_rel_name must be the name of a noisy relation.

Reply:
I had neglected the clean relation case, I added it.